### PR TITLE
[ValSet-Pref] UnDelegate tokens from val-set using MsgUnDelegateFromValidatorSet

### PR DESCRIPTION
**Pull request by stackman27**
Part of : https://github.com/osmosis-labs/osmosis/issues/2579
Merge https://github.com/osmosis-labs/osmosis/pull/3260 first 

## What is the purpose of the change

`UndelegateFromValidatorSet` undelegates {coin} amount from the validator set.
For ex: userA has staked 10tokens with weight {Val->0.5, ValB->0.3, ValC->0.2}
undelegate 6osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}
our undelegate logic would attempt to undelegate 3osmo from A , 1.8osmo from B, 1.2osmo from C
## Testing and Verifying

Added tests for TestUnDelegateFromValidatorSet

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/valset-pref/validator_set.go [hunk 1]
```diff
@@ -29,6 +29,99 @@ func (k Keeper) SetupValidatorSetPreference(ctx sdk.Context, delegator string, p
...
+
+	// the total amount the user wants to undelegate
+	tokenAmt := sdk.NewDec(coin.Amount.Int64())
+
+	totalAmountFromWeights := sdk.NewDec(0)
+	for _, val := range existingSet.Preferences {
+		// Calculate the amount to undelegate based on the existing weights
+		amountToUnDelegate := val.Weight.Mul(tokenAmt)
+
+		// ValidateValidator gurantees that this exist
```
Comment (czarcas7ic) [inline x/valset-pref/validator_set.go:94]:
```suggestion
		// ValidateValidator guarantees that this exist
```

File: x/valset-pref/validator_set.go [hunk 2]
```diff
@@ -29,6 +29,99 @@ func (k Keeper) SetupValidatorSetPreference(ctx sdk.Context, delegator string, p
...
+		if err != nil {
+			return err
+		}
+
+		totalAmountFromWeights = totalAmountFromWeights.Add(amountToUnDelegate)
+	}
+
+	if !totalAmountFromWeights.Equal(tokenAmt) {
+		return fmt.Errorf("The undelegate total donot add up with the amount calculated from weights expected %s got %s", tokenAmt, totalAmountFromWeights)
+	}
```
Comment (czarcas7ic) [inline x/valset-pref/validator_set.go:120]: To be clear, this error check is nothing but informative right? Bc regardless if it matches or not, the undelegations will have taken place

**Review by czarcas7ic — COMMENTED** [review]
This seems good to me but this includes changes from the other PR which do have some desired changes. Will approve once that once is merged into this

**Review by mattverse — COMMENTED** [review]
Consider merging this PR with the DelegateFromValidatorSet PR.

With this Pr also including the changes for the DelegateFromValidatorSet message, I think it's most likely that we make a change over there and then we do a double review or you need to resolve all the merge conflicts in the future. Also, I personally think having Delegate and Undelgate Message in the same PR makes sense!

File: x/valset-pref/validator_set.go (hunk 2 above)
Comment (stackman27) [reply x/valset-pref/validator_set.go:120]: ah yes! i'll put this logic before `undelegate`, the only downside would be to run the loop twice

**Comment by stackman27** [comment]
Closing this PR in favor of adding Delegate and Undelegate to the same PR https://github.com/osmosis-labs/osmosis/pull/3260

**Pull request closed without merging**